crates/auth/src/types.rs

```diff
@@ -0,0 +1,166 @@
+// SPDX-FileCopyrightText: 2022 Profian Inc. <opensource@profian.com>
+// SPDX-License-Identifier: Apache-2.0
+use std::fmt;
+
+use axum::{
+    async_trait,
+    extract::{
+        rejection::TypedHeaderRejectionReason, Extension, FromRequest, RequestParts, TypedHeader,
+    },
+    response::{IntoResponse, Redirect, Response},
+};
+use http::header::{self, USER_AGENT};
+use oauth2::AccessToken;
+use rand::rngs::OsRng;
+use rsa::{PaddingScheme, PublicKey, RsaPrivateKey};
+use serde::{Deserialize, Serialize};
+
+use crate::{error::Error, COOKIE_NAME, GITHUB};
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum AuthType {
+    GitHub,
+}
+
+impl fmt::Display for AuthType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                AuthType::GitHub => "GitHub.com",
+            }
+        )
+    }
+}
+
+#[test]
+fn auth_type_display() {
+    assert_eq!(format!("{}", AuthType::GitHub), "GitHub.com");
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct User {
+    pub auth_type: AuthType,
+    pub token: AccessToken,
+}
+
+impl User {
+    /// Decrypt an untrusted user token to guarantee it was not modified by the user.
+    pub fn from_encrypted_str(string: &str, key: &RsaPrivateKey) -> Result<Self, Error> {
+        let bytes = base64::decode(string).map_err(Error::TokenDecode)?;
+        let bytes = key
+            .decrypt(PaddingScheme::new_pkcs1v15_encrypt(), &bytes)
+            .map_err(Error::TokenDecrypt)?;
+        bincode::deserialize(&bytes).map_err(Error::TokenDeserialize)
+    }
+
+    /// Encrypt the session so that it can be securely stored by the user.
+    pub fn encrypt(&self, key: &RsaPrivateKey) -> Result<String, Error> {
+        let bytes = bincode::serialize(self).map_err(Error::TokenSerialize)?;
+        let bytes = key
+            .encrypt(
+                &mut OsRng,
+                PaddingScheme::new_pkcs1v15_encrypt(),
+                &bytes[..],
+            )
+            .map_err(Error::TokenEncrypt)?;
+        Ok(base64::encode(bytes))
+    }
+
+    /// Validate if a session should be considered active.
+    pub async fn validate(&self) -> Result<String, Error> {
+        match self.auth_type {
+            AuthType::GitHub => {
+                let client = reqwest::Client::new();
+
+                #[derive(Deserialize)]
+                struct GitHubUser {
+                    login: String,
+                }
+
+                #[derive(Deserialize)]
+                struct GitHubError {
+                    message: String,
+                }
+
+                let body = client
+                    .get("https://api.github.com/user")
+                    .header(USER_AGENT, "drawbridge")
+                    .bearer_auth(self.token.secret())
+                    .send()
+                    .await
+                    .map_err(Error::Request)?
+                    .text()
```

Comment:
We probably want to use https://docs.serde.rs/serde_json/fn.from_reader.html on the body directly without calling `text()`

Reply:
I have to retry if it fails to deserialize to get my GitHubError out